### ISSUE TITLE
[QUESTION]: Downloading 3rd party assets during build and package

### ISSUE THREAD

**Issue by onurbingol** [NVIDIA-Omniverse/kit-app-template#39]
### Question

Is there a way to download 3rd party assets, such as USD files required for custom extensions, from S3/Azure buckets while executing `repo.sh build`? I am actively using `pip.toml` to download and package python packages required for the extensions I am developing and I was wondering if a similar method would be available for S3/Azure.

If not available, I'd be very happy if you would consider adding such a feature.

**Comment by SchultzC** [comment]
This is not something that we have streamlined via tooling.  

If this is something that you feel would meaningfully improve your development workflow let's get a feature request going and dig into more details about how this would work and how it would be used (and the value it would bring).

**Comment by onurbingol** [comment]
Initially, the number of USD assets were small and I stored them on Git under an extension. We have been adding more assets every day (models, materials, textures, mdl files, etc.) and I thought that it would be good to have an option like an asset prebundle, very much like pip prebundle. So that, I can store them separately on a blob storage and download whenever I am building the package. I'd probably need to download them during development as well, but I think development vs. production difference can be handled programmatically within the extension which is consuming the assets.

**Comment by SchultzC** [comment]
In your application workflow is it required to include the asset(s) as part of the package?

Could you download the data as needed at run time.  For example, on load of the extension or upon user request/action?

**Comment by onurbingol** [comment]
> In your application workflow is it required to include the asset(s) as part of the package?
>
> Could you download the data as needed at run time. For example, on load of the extension or upon user request/action?

One issue preventing implementation of this suggestion is our use of private clouds. Our CI has download access, but not all users have access to the cloud. We could implement this for developers, but we would prefer to package everything for regular users. Since all the packages are versioned on our side, we would know what assets are used in a release.

**Comment by SchultzC** [comment]
I see. Thank you for your patience on this. You can download assets as part of your build by running a custom script after the main build completes. In your top level `repo.toml` file you can define a post build step, for example:

```
[repo_build.post_build]
commands = [
["{path}/my_downloader_script.{shell_ext}", "-{flag/arg}"]
]
```

Keep in mind that these commands are not shell commands; they're lists of arguments passed to Python's subprocess. You can test them by calling subprocess.Popen([...]) in Python. Once your script downloads the assets into your extension folder, the standard packaging tool will pick them up in the final artifacts.

If this meets your needs, great. Otherwise, we can discuss adding a mechanism for large third-party assets as a feature request.

**Comment by onurbingol** [comment]
I have started using a kit app with a single custom python extension dependency. This custom extension is designed to download the assets I need from various sources to the folders in the kit-app-template directory structure. I control all the necessary operations using the Kit variables and command-line arguments. I also use a pip prebundle configuration to install the required Python wheel dependencies. It sounds like overkill to implement it in this way, but considering the time passed between December 2nd, 2024 and April 11th, 2025, I had to invent something. 

Even though I don't need it anymore, using your suggestion might also work for kit-app-template users' similar use cases.

**Comment by SchultzC** [comment]
You're right to point out the time frame from the opening of the questions to an answer.  It took too long.  I will find a way to reduce turn around time, especially for Question type Issues.

Thanks for sharing your workaround, I'm glad you found a way that works for you.  

We appreciate your persistence.
